src/ray/core_worker/core_worker.cc

```diff
@@ -560,25 +585,45 @@ bool CoreWorker::AddActorHandle(std::unique_ptr<ActorHandle> actor_handle) {
     // Register a callback to handle actor notifications.
     auto actor_notification_callback = [this](const ActorID &actor_id,
                                               const gcs::ActorTableData &actor_data) {
-      if (actor_data.state() == gcs::ActorTableData::RECONSTRUCTING) {
-        auto it = actor_handles_.find(actor_id);
-        RAY_CHECK(it != actor_handles_.end());
-        if (it->second->IsDirectCallActor()) {
-          // We have to reset the actor handle since the next instance of the
-          // actor will not have the last sequence number that we sent.
-          // TODO: Remove the check for direct calls. We do not reset for the
-          // raylet codepath because it tries to replay all tasks since the
-          // last actor checkpoint.
-          it->second->Reset();
+      auto it = actor_handles_.find(actor_id);
+      RAY_CHECK(it != actor_handles_.end());
+      if (actor_data.state() == gcs::ActorTableData::ALIVE) {
+        it->second->Connect(actor_data, client_call_manager_);
+        direct_actor_submitter_->SendPendingTasks(actor_id, it->second->RpcClient());
+      } else {
+        // We have to reset the actor handle since the next instance of the
+        // actor will not have the last sequence number that we sent.
+        // TODO: Remove the flag for direct calls. We do not reset for the
+        // raylet codepath because it tries to replay all tasks since the last
+        // actor checkpoint.
+        it->second->Reset(actor_data,
+                          /*reset_task_counter=*/it->second->IsDirectCallActor());
+        direct_actor_submitter_->FailPendingTasks(actor_id);
+
+        if (actor_data.state() == gcs::ActorTableData::RECONSTRUCTING) {
```

Comment:
Can this be in a helper function?

Reply:
Actually it would be great if this were pulled out into a class like ActorManager. It would also then become possible to unit test handling of actor restarts.

Reply:
Which part specifically did you want to pull out?